sysdump/sysdump.go

```diff
@@ -0,0 +1,825 @@
+				p, err := c.k.ListPods(ctx, c.o.CiliumOperatorNamespace, metav1.ListOptions{
+					LabelSelector: c.o.CiliumOperatorLabelSelector,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to get logs from Cilium operator pods")
+				}
+				if err := c.collectLogs(ctx, c.o.Logger, filterPods(p, nodeList), sinceTime, c.o.LogsLimitBytes, absoluteTempPath); err != nil {
+					return fmt.Errorf("failed to collect logs from Cilium operator pods")
+				}
+				return nil
+			},
+			// Get logs from Hubble.
+			func() error {
+				c.o.Logger.Info("Collecting logs from Hubble pods")
+				defer c.o.Logger.Debug("Finished collecting logs from Hubble pods")
+				p, err := c.k.ListPods(ctx, c.o.HubbleNamespace, metav1.ListOptions{
+					LabelSelector: c.o.HubbleLabelSelector,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to get logs from Hubble pods")
+				}
+				if err := c.collectLogs(ctx, c.o.Logger, filterPods(p, nodeList), sinceTime, c.o.LogsLimitBytes, absoluteTempPath); err != nil {
+					return fmt.Errorf("failed to collect logs from Hubble pods")
+				}
+				return nil
+			},
+			// Get logs from Hubble Relay.
+			func() error {
+				c.o.Logger.Info("Collecting logs from Hubble Relay pods")
+				defer c.o.Logger.Debug("Finished collecting logs from Hubble Relay pods")
+				p, err := c.k.ListPods(ctx, c.o.HubbleRelayNamespace, metav1.ListOptions{
+					LabelSelector: c.o.HubbleRelayLabelSelector,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to get logs from Hubble Relay pods")
+				}
+				if err := c.collectLogs(ctx, c.o.Logger, filterPods(p, nodeList), sinceTime, c.o.LogsLimitBytes, absoluteTempPath); err != nil {
+					return fmt.Errorf("failed to collect logs from Hubble Relay pods")
+				}
+				return nil
+			},
+			// Get logs from Hubble UI.
+			func() error {
+				c.o.Logger.Info("Collecting logs from Hubble UI pods")
+				defer c.o.Logger.Debug("Finished collecting logs from Hubble UI pods")
+				p, err := c.k.ListPods(ctx, c.o.HubbleNamespace, metav1.ListOptions{
+					LabelSelector: c.o.HubbleLabelSelector,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to get logs from Hubble UI pods")
+				}
+				if err := c.collectLogs(ctx, c.o.Logger, filterPods(p, nodeList), sinceTime, c.o.LogsLimitBytes, absoluteTempPath); err != nil {
+					return fmt.Errorf("failed to collect logs from Hubble UI pods")
+				}
+				return nil
+			},
+		}...)
+	}
+
+	// Run the tasks.
+	errCh := make(chan error)
+	wgdCh := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(len(tasks))
+	for _, f := range tasks {
+		go func(f func() error) {
+			defer wg.Done()
+			if err := f(); err != nil {
+				errCh <- err
+			}
+		}(f)
+	}
+	go func() {
+		wg.Wait()
+		close(wgdCh)
+	}()
+
+	// Wait for the tasks to finish.
+	var merr error
+loop:
+	for {
+		select {
+		case <-wgdCh:
+			break loop
+		case err := <-errCh:
+			merr = multierror.Append(merr, err)
+		}
+	}
```

Comment:
Have you considered using [`cilium/workerpool`](https://github.com/cilium/workerpool) here? I think you could spare some lines of code and gain in readability here.

Reply:
Interesting, I wasn't aware of this! Let me 🔍!

Reply:
Done — though a `WaitGroup` is still in place to wait for all sub-tasks to be submitted before calling `Drain()`. Thank you for your help on this @rolinh. 💯